tests/discrete_demography_roundtrips.cpp

```diff
@@ -0,0 +1,367 @@
+
+{
+    using config_fillter_function_type = std::function<void(
+        const metadata_vector&, fwdpy11::PopulationConfiguration&)>;
+    using metadata_update_function_type = std::function<void(
+        const fwdpy11::PopulationConfiguration&, metadata_vector&)>;
+    static_assert(
+        std::is_convertible<config_filler_function,
+                            config_fillter_function_type>::value,
+        "Invalid function signature for filling PopulationConfiguration "
+        "object.");
+    static_assert(std::is_convertible<metadata_update_function,
+                                      metadata_update_function_type>::value,
+                  "Invalid function signature for updating metadata");
+    // It is O(N) to update config, but O(1) to find out
+    // if we don't need to, so checking is a win even1
+    // if config changes are relatively common
+    bool config_changes = demography.next_config_event_at(parental_generation);
+    if (config_changes)
+        {
+            filler(metadata, config);
+            demography.change_config(parental_generation, rng, config);
+            update_metadata(config, metadata);
+        }
+    auto p = demography.change_params(parental_generation);
+    auto m = demography.change_migration_rates(parental_generation);
+    demography.apply_growth_rates(parental_generation + 1);
+    return { config_changes, p, m };
+}
+
+void
+setup_initial_metadata(fwdpy11::DiploidPopulation& pop,
+                       fwdpy11::SoftSelection& demography)
+//NOTE: this is a total failure if pop is already
+//set up ok.  It will also fail if we call
+//an evolve function more than once.
+{
+    auto popsizes = demography.current_deme_sizes();
+    if (popsizes.empty())
+        {
+            throw std::invalid_argument("deme sizes are empty");
+        }
+    std::uint32_t ttl = 0;
+    for (std::size_t deme = 0; deme < popsizes.size(); ++deme)
+        {
+            for (std::uint32_t i = 0; i < popsizes[deme]; ++i)
+                {
+                    pop.diploid_metadata[i].deme = deme;
+                }
+            ttl += popsizes[deme];
+        }
+    if (ttl != pop.N)
+        {
+            throw std::invalid_argument(
+                "initial states do not equal population size");
+        }
+}
+
+void
+SoftSelection_roundtrip(const fwdpy11::GSLrng_t& rng,
+                        fwdpy11::DiploidPopulation& pop,
+                        fwdpy11::SoftSelection& demography,
+                        std::uint32_t ngens)
+{
+    demography.finalize_callbacks();
+    setup_initial_metadata(pop, demography);
+    auto offspring_metadata(pop.diploid_metadata);
+    auto offspring(pop.diploids);
+    fwdpy11::PopulationConfiguration config;
+    std::vector<fitness_lookup> lookups;
+    auto migtable = demography.migration_lookup_table();
+    std::size_t pdeme1, pdeme2;
+    for (std::uint32_t gen = 0; gen < ngens; ++gen, ++pop.generation)
+        {
```

Comment:
Nearly all of the ops in this look can be abstracted back to the DiscreteDemography classes.  Several of the functions called here can likely be made private/protected members of those classes as well.